fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/expressions/SelectExpression.java

```diff
@@ -342,15 +339,107 @@ public Iterable<MatchInfo> subsumedBy(@Nonnull final RelationalExpression candid
             }
         }
 
+        //
+        // Go through each predicate on the query side in an attempt to find predicates on the candidate side that
+        // can subsume the current predicate. For instance, `x < 5` on the query side can be subsumed by a placeholder
+        // for x in the candidate. A predicate `x < 5' is not subsumed by 'x = 10' as the set of records retained by
+        // applying 'x = 10' does not contain all of those retained after applying 'x < 5'.
+        // Predicates on the query side can at least be subsumed by a tautology 'true' which should always be possible.
+        // Whenever the pairing is lossy, for instance `x < 5' is subsumed by 'x < 10', the query side predicate needs
+        // to be compensated if and when the match is realized. Compensating a predicate usually means its reapplication
+        // as a residual filter. If a predicate is matched to a placeholder, the match is considered lossless as the
+        // placeholder can be morphed into the right lossless pairing.
+        //
+        // Join predicates: Join predicates are matched the same way as local predicates with a complication.
+        // If a join come across a join predicate we need to consider the potential other application of this predicate
+        // in another match and/or match candidate.
+        //
+        // Example:
+        //
+        // SELECT *
+        // FROM R r, S s
+        // WHERE r.a < 5 AND s.b = 10 AND r.x = s.y
+        //
+        // The predicate 'r.x = r.y' can be used as a predicate for matching an index on R(x, a) or for
+        // matching an index on S(b, y). In fact the predicate needs to be shared in some way such that the planner
+        // can later on make the right decision based on cost, etc.
+        //
+        // The way this is implemented is to create two matches one where the predicate is repossessed to the match
+        // at hand. When we match R(x, a) we repossess r.x = r.y to be subsumed by r.x = ? on
+        // the candidate side. Vica versa, when we match S(b, y) we repossess s.y = r.x to be subsumed by s.y = ? on the
+        // candidate side.
+        //
+        // Using this approach we create a problem that these two matches cannot coexist in a way that they cannot
+        // be realized, that is planned together at all as both matches provide the other's placeholder value. In fact,
+        // we have forced the match to (if it were to be planned) become the inner of a join. It would be beneficial,
+        // however, to also create a version of the match that does not consider the join predicate at all.
```

Comment:
> to also create a version of the match that does not consider the join predicate at all.

so it will be added as a compensation on top right?

i.e. in your example:
`FlatMap(IScan(S, b, [10, 10]), IScan(R, s, [-inf, 5[))`
Then, the join will be applied as a residual:
`LogicalFilterExpression(Equals(x,y), (FlatMap(IScan(S, b, [10, 10]), IScan(R, s, [-inf, 5[)))`.
right?

If so, does it make sense to have this alternative because it looks to me quite expensive, so it might make sense to avoid it even without a cost model, but honestly I haven't thought about it deep enough.

Reply:
You need to always create a match that does not repossess a join predicate, because creating that match with the predicate will make the match dependent on the other correlations in that join predicate. That's bad.

A realized match that rejects such a predicate will not compensate for it either, since some other match may compensate for it (or use it as an index sargable). If no other match wants it, the realized match using the join predicate will probably win out the costing, but if not, the code in `SelectDataAccessRule` compensates for all that's left for the expression. I hope that's helpful.

Reply:
Yes, it is, thanks.